File: functions/src/lib/name_key.test.ts
import { readFileSync } from "fs";
import { resolve } from "path";
import { describe, expect, it } from "vitest";
import { nameKeyOf, normalizeName } from "./name_key";

/**
 * Shared fixture read by BOTH the TypeScript and the Dart test suites (Issue #88),
 * so the two `normalizeName` implementations cannot drift apart:
 *
 *   - TS   : functions/src/lib/name_key.ts   <- this file
 *   - Dart : lib/core/utils/name_key.dart    <- test/core/name_key_test.dart
 *
 * Path is resolved from this file (functions/src/lib) up to the repository root.
 */
const FIXTURE_PATH = resolve(
  __dirname,
  "../../../test/fixtures/name_normalization_cases.json",
);

type NormalizationCase = {
  description: string;
  input: string;
  expected: string;
};

const fixture: { cases: NormalizationCase[] } = JSON.parse(
  readFileSync(FIXTURE_PATH, "utf8"),
);

describe("normalizeName (shared fixture with the Dart implementation)", () => {
  it("loads the shared fixture", () => {
    expect(fixture.cases.length).toBeGreaterThan(0);
  });

  it.each(fixture.cases)("$description", ({ input, expected }) => {
    expect(normalizeName(input)).toBe(expected);
  });
});

describe("nameKeyOf", () => {
  it("produces the same key for equivalent display variants", () => {
    expect(nameKeyOf("  牛乳  ")).toBe(nameKeyOf("牛乳"));
    expect(nameKeyOf("Milk")).toBe(nameKeyOf("milk"));
    expect(nameKeyOf("Ｍｉｌｋ")).toBe(nameKeyOf("milk"));
    expect(nameKeyOf("カット   トマト")).toBe(nameKeyOf("カット トマト"));
  });

  it("produces different keys for different products", () => {
    expect(nameKeyOf("牛乳")).not.toBe(nameKeyOf("豆乳"));
  });

  it("produces a Firestore-document-id-safe key (no '/' characters)", () => {
    const key = nameKeyOf("牛乳/豆乳 100% こだわり");
    expect(key).not.toContain("/");
    expect(key.length).toBeGreaterThan(0);
  });

  it("is deterministic", () => {
    expect(nameKeyOf("にんじん")).toBe(nameKeyOf("にんじん"));
  });
});

File: test/core/name_key_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:shopping_list_app/core/utils/name_key.dart';

/// Dart / TypeScript 双方のテストが読む共通フィクスチャ（#88）。
///
/// TS 側は `functions/src/lib/name_key.test.ts` が同じファイルを読む。
/// 片側だけを更新すると必ずもう一方のテストが落ちるため、両実装の乖離を防げる。
const _fixturePath = 'test/fixtures/name_normalization_cases.json';

void main() {
  final fixture =
      jsonDecode(File(_fixturePath).readAsStringSync()) as Map<String, dynamic>;
  final cases = (fixture['cases'] as List<dynamic>)
      .cast<Map<String, dynamic>>();

  group('normalizeName（TS 実装との共通フィクスチャ）', () {
    test('フィクスチャが読み込めている', () {
      expect(cases, isNotEmpty);
    });

    for (final testCase in cases) {
      test(testCase['description'] as String, () {
        expect(
          normalizeName(testCase['input'] as String),
          testCase['expected'] as String,
        );
      });
    }
  });

  group('normalizeName（同一視の性質）', () {
    test('トリム後に「牛乳」と「牛乳 」が同一正規化になる', () {
      expect(normalizeName('牛乳'), equals(normalizeName('牛乳 ')));
    });

    test('全角英字と半角英字が同一正規化になる（#88 の再現ケース）', () {
      expect(normalizeName('ＡＢＣ牛乳'), equals(normalizeName('ABC牛乳')));
    });

    test('半角カナと全角カナが同一正規化になる', () {
      expect(normalizeName('ﾐﾙｸ'), equals(normalizeName('ミルク')));
    });

    test('全角スペース区切りと半角スペース区切りが同一正規化になる', () {
      expect(normalizeName('牛乳　パン'), equals(normalizeName('牛乳 パン')));
    });

    test('異なる商品は別の正規化結果になる', () {
      expect(normalizeName('牛乳'), isNot(equals(normalizeName('豆乳'))));
    });
  });
}
